pkg/dynamicaccess/history.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dynamicaccess
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/ethersphere/bee/v2/pkg/file"
+	"github.com/ethersphere/bee/v2/pkg/manifest"
+	"github.com/ethersphere/bee/v2/pkg/manifest/mantaray"
+	"github.com/ethersphere/bee/v2/pkg/swarm"
+)
+
+type History interface {
+	Add(ctx context.Context, ref swarm.Address, timestamp *int64, metadata *map[string]string) error
+	Lookup(ctx context.Context, timestamp int64) (manifest.Entry, error)
+	Store(ctx context.Context) (swarm.Address, error)
+}
+
+var _ History = (*HistoryStruct)(nil)
+
+var (
+	ErrEndIteration   = errors.New("end iteration")
+	ErrUnexpectedType = errors.New("unexpected type")
+)
+
+type HistoryStruct struct {
+	manifest *manifest.MantarayManifest
```

Comment:
What to return in the case of a simple manifest ? If Root() is added to the manifest interface then the return type cannot be a Mantaray Node.

Reply:
Since the history implementation depends on the mantaray implementation to walk and take the timestamps from them, the simple manifest cannot be used for history. To avoid exporting manifest.MantarayManifest, just to get the root node, an unexported local interface that clearly states what is used, alongside with the correct constructor can be used, where the constructor is adjusted to return the concrete type and not the interface. A simple example that would be:

```go
+type manifestInterface interface {
+       manifest.Interface
+       Root() *mantaray.Node
+}
+
 var (
        ErrEndIteration   = errors.New("end iteration")
        ErrUnexpectedType = errors.New("unexpected type")
 )
 
 type HistoryStruct struct {
-       manifest *manifest.MantarayManifest
+       manifest manifestInterface
        ls       file.LoadSaver
 }
 
 func NewHistory(ls file.LoadSaver) (*HistoryStruct, error) {
-       m, err := manifest.NewDefaultManifest(ls, false)
+       m, err := manifest.NewMantarayManifest(ls, false)
        if err != nil {
-               return nil, fmt.Errorf("failed to create default manifest: %w", err)
+               return nil, fmt.Errorf("failed to create materay manifest: %w", err)
        }
 
-       mm, ok := m.(*manifest.MantarayManifest)
+       mm, ok := m.(manifestInterface)
        if !ok {
                return nil, fmt.Errorf("%w: expected MantarayManifest, got %T", ErrUnexpectedType, m)
        }
@@ -51,12 +56,12 @@ func NewHistory(ls file.LoadSaver) (*HistoryStruct, error) {
 }
 
 func NewHistoryReference(ls file.LoadSaver, ref swarm.Address) (*HistoryStruct, error) {
-       m, err := manifest.NewDefaultManifestReference(ref, ls)
+       m, err := manifest.NewMantarayManifestReference(ref, ls)
        if err != nil {
-               return nil, fmt.Errorf("failed to create default manifest: %w", err)
+               return nil, fmt.Errorf("failed to create materay manifest: %w", err)
        }
 
-       mm, ok := m.(*manifest.MantarayManifest)
+       mm, ok := m.(manifestInterface)
        if !ok {
                return nil, fmt.Errorf("%w: expected MantarayManifest, got %T", ErrUnexpectedType, m)
        }
```

This would protect the runtime from an unexpected error on default manifest change because the mantaray specific constructor are used.

There could be different approaches to avoid excessive exports and even to avoid adding the Root() method by moving the lookupNode method logic to the mantaray package (it really is a function only on mantaray entities), or another approach to have additional constructors that return trie root as well, and the current mantaray manifest constructors use them just to ignore the trie root node.

I just want to illustrate that there are multiple solutions that can have less impact on the existing code and I think that exporting mantarayManifest type just to get the root node is not one of the better approaches.

Reply:
fixed it with the 'manifestInterface' solution